File: sections/35_math_mixed_contest/problems/h_gcd_one_subsets/solution.cpp
#include <bits/stdc++.h>
using namespace std;

const int MOD = 1'000'000'007;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin >> n;
    vector<int> values(n);
    int maximum = 0;
    for (int& value : values) {
        cin >> value;
        maximum = max(maximum, value);
    }

    vector<int> frequency(maximum + 1);
    for (int value : values)
        ++frequency[value];

    vector<int> powers(n + 1, 1);
    for (int i = 1; i <= n; ++i)
        powers[i] = 2LL * powers[i - 1] % MOD;

    vector<int> exact(maximum + 1);
    for (int divisor = maximum; divisor >= 1; --divisor) {
        int divisible_count = 0;
        long long remove = 0;
        for (int multiple = divisor; multiple <= maximum; multiple += divisor) {
            divisible_count += frequency[multiple];
            if (multiple != divisor)
                remove += exact[multiple];
        }
        exact[divisor] = (powers[divisible_count] - 1 - remove) % MOD;
        if (exact[divisor] < 0)
            exact[divisor] += MOD;
    }

    cout << exact[1] << '\n';
}

File: sections/35_math_mixed_contest/problems/h_gcd_one_subsets/solution.py
import sys


MOD = 1_000_000_007


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    n = data[0]
    values = data[1:]
    maximum = max(values)

    frequency = [0] * (maximum + 1)
    for value in values:
        frequency[value] += 1

    powers = [1] * (n + 1)
    for i in range(1, n + 1):
        powers[i] = 2 * powers[i - 1] % MOD

    exact = [0] * (maximum + 1)
    for divisor in range(maximum, 0, -1):
        divisible_count = 0
        remove = 0
        for multiple in range(divisor, maximum + 1, divisor):
            divisible_count += frequency[multiple]
            if multiple != divisor:
                remove += exact[multiple]
        exact[divisor] = (powers[divisible_count] - 1 - remove) % MOD

    print(exact[1])


if __name__ == "__main__":
    main()
